src/controllers/document.download.controller.ts

```diff
@@ -0,0 +1,103 @@
+import { Session } from "ch-node-session-handler";
+import { Request, Response } from "express";
+import { DownloadErrorMessages, HttpStatusCodes } from "../model/error.messages";
+import { Templates } from "../model/template.paths";
+import {
+  Download,
+  HEADER_CONTENT_DISPOSITION,
+  HEADER_CONTENT_LENGTH,
+  HEADER_CONTENT_TYPE,
+} from "../modules/sdk/objections";
+import { downloadAttachment } from "../services/objection.service";
+import logger from "../utils/logger";
+import { DOWNLOAD_FILENAME_PREFIX } from "../utils/properties";
+
+/**
+ * Handle GET request - do the download
+ * @param {Request} req the http request object
+ * @param {Response} res the http response object
+ */
+export const get = async (req: Request, res: Response) => {
+  const downloadUrl = req.url;
+  logger.infoRequest(req, `Attempting download of ${downloadUrl}`);
+
+  const session: Session = req.session as Session;
+
+  try {
+    // get the download from api
+    const download: Download = await downloadAttachment(downloadUrl, session);
+
+    // set the file info in the response headers
+    setResponseHeaders(res, download);
+
+    // pipe/stream the data to the client (browser) through the response object
+    logger.debugRequest(req, `About to pipe download to browser from  ${downloadUrl}`);
+    return download.data.pipe(res);
+
+  } catch (e) {
+    logger.errorRequest(req, `download.attachment.controller - ${JSON.stringify(e)}`);
+    const status: number = (e.status > 0) ? e.status : 500;
+    res.status(status);
+    return showErrorPage(status, res);
+  }
+};
+
+/**
+ * Set the response headers for the download using the headers returned from the API that contain
+ * file information
+ * @param {Response} res the http response object
+ * @param {Download} download the download data returned from API
+ */
+const setResponseHeaders = (res: Response, download: Download) => {
+  res.setHeader(HEADER_CONTENT_DISPOSITION, prefixFilename(download.headers[HEADER_CONTENT_DISPOSITION]));
+  res.setHeader(HEADER_CONTENT_TYPE, download.headers[HEADER_CONTENT_TYPE]);
+  res.setHeader(HEADER_CONTENT_LENGTH, download.headers[HEADER_CONTENT_LENGTH]);
+
+  logger.debug("Returning response with headers " + JSON.stringify(res.getHeaders()));
+};
+
+/**
+ * Prefix filename with a value from config
+ * @param {string} contentDisposition
+ * @returns {string} the contentDisposition string with filename prefix applied
+ */
+const prefixFilename = (contentDisposition: string ): string => {
+  const FILENAME_ID = "filename=\"";
+  return contentDisposition.replace(FILENAME_ID, FILENAME_ID + DOWNLOAD_FILENAME_PREFIX);
+};
+
+/**
+ * Show the correct error page to the user
+ * @param {number} httpStatus
+ * @param {Response} res
+ */
+const showErrorPage = (httpStatus: number, res: Response) => {
+  const { UNAUTHORIZED, FORBIDDEN, NOT_FOUND } = HttpStatusCodes;
+  switch (httpStatus) {
+    case UNAUTHORIZED: {
```

Comment:
I've just had the thought, from a security perspective if they are unauthorized or forbidden, we should probably act as if the file wasnt found.
This way a malicious user can use the "Not authorized" response to know that a file of given id exists.

Reply:
I think Sarah has already discussed with the business what messages will be shown - she's created a story for the forbidden one - https://companieshouse.atlassian.net/browse/OBJ-190. I can let her know.

Reply:
Mayb be worth double checking with Steve or Kev as user requirements may clash with security recommendations here

Reply:
updated, does this mean we'll have to change the API to return 404 as well

Reply:
I suppose we should, though as the api is internal that could be relatively low priority?

Reply:
👍 should create a tech story to do this, so it's not forgotten. I'll do this if you like.